### PR TITLE
Speedup futility_move_count

### PR DESCRIPTION
**Pull request by cj5716**
Change from ```return improving ? (3 + depth * depth)
                     : (3 + depth * depth) / 2;```
to `return (3 + depth * depth) >> (improving ? 0 : 1);`
Credit is shared with @maximmasiutin for suggesting to use `>>` function.

Passed STC: 
LLR: 2.94 (-2.94,2.94) <0.00,2.00>
Total: 107632 W: 28645 L: 28248 D: 50739
Ptnml(0-2): 307, 11134, 30561, 11483, 331
https://tests.stockfishchess.org/tests/view/6423a012db43ab2ba6f95e05

No functional change

### CODE REVIEW

File: src/search.cpp [hunk 1]
```diff
@@ -76,8 +76,7 @@ namespace {
   }
 
   constexpr int futility_move_count(bool improving, Depth depth) {
-    return improving ? (3 + depth * depth)
-                     : (3 + depth * depth) / 2;
+    return (3 + depth * depth) >> (improving ? 0 : 1);
```
Comment (Disservin) [inline src/search.cpp:79]: I find the ternary for boolean’s pretty ugly I would suggest to simply use !improving though you might have to rerun this test

**Comment by vondele** [comment]
can you also run the LTC test, so we're sure this is not a fluke? A good compiler should optimize this.

File: src/search.cpp (hunk 1 above)
Comment (Disservin) [reply src/search.cpp:79]: Okay then I think it’s fluke, like vondele mentioned.

**Comment by dubslow** [comment]
A should-be-identical respelling failed STC: https://tests.stockfishchess.org/tests/view/6424fe11db43ab2ba6f99d17

LTC is a fairly poor way to measure speedups, being both resource intensive and considerably less sensitive to nps than STC. Instead I suggest that both this version and its respelling be tested again at STC.

**Comment by mstembera** [comment]
Godbolt shows the assembly produced is actually a bit different.  https://godbolt.org/z/fxvTddej9

**Comment by maximmasiutin** [comment]
@mstembera - interesting. How to ask the site to choose compilers? clang 16 with "-O3" produces the following results:

```
0000000000000650  futility_move_count:
 650:   89 f9                   mov    %edi,%ecx
 652:   0f af f6                imul   %esi,%esi
 655:   8d 46 03                lea    0x3(%rsi),%eax
 658:   80 f1 01                xor    $0x1,%cl
 65b:   d3 e8                   shr    %cl,%eax
 65d:   c3                      ret


0000000000000660 futility_move_count2
 660:   89 f9                   mov    %edi,%ecx
 662:   0f af f6                imul   %esi,%esi
 665:   8d 46 03                lea    0x3(%rsi),%eax
 668:   80 f1 01                xor    $0x1,%cl
 66b:   d3 e8                   shr    %cl,%eax
 66d:   c3                      ret


0000000000000670 futility_move_count3:
 670:   89 f9                   mov    %edi,%ecx
 672:   0f af f6                imul   %esi,%esi
 675:   8d 46 03                lea    0x3(%rsi),%eax
 678:   80 f1 01                xor    $0x1,%cl
 67b:   d3 e8                   shr    %cl,%eax
 67d:   c3                      ret
```

The source code is the following:
```
int futility_move_count(bool improving, int depth) {
    return improving ? (3 + depth * depth)
                     : (3 + depth * depth) / 2;
}

int futility_move_count2(bool improving, int depth) {
    return (3 + depth * depth) >> (improving ? 0 : 1);
}

int futility_move_count3(bool improving, int depth) {
    return (3 + depth * depth) >> (!improving);
}

int main(int argc, char *argv[])
{
  return
   futility_move_count(argc, argc) +
   futility_move_count2(argc, argc) +
   futility_move_count3(argc, argc);
}

```

**Comment by maximmasiutin** [comment]
But gcc 12.2.0 produces the following:

```
00000000000006d0 futility_move_count:
 6d0:   f3 0f 1e fa             endbr64
 6d4:   0f af f6                imul   %esi,%esi
 6d7:   8d 46 03                lea    0x3(%rsi),%eax
 6da:   89 c2                   mov    %eax,%edx
 6dc:   d1 fa                   sar    %edx
 6de:   40 84 ff                test   %dil,%dil
 6e1:   0f 44 c2                cmove  %edx,%eax
 6e4:   c3                      ret

00000000000006f0 futility_move_count2:
 6f0:   f3 0f 1e fa             endbr64
 6f4:   0f af f6                imul   %esi,%esi
 6f7:   89 f9                   mov    %edi,%ecx
 6f9:   83 f1 01                xor    $0x1,%ecx
 6fc:   8d 46 03                lea    0x3(%rsi),%eax
 6ff:   d3 f8                   sar    %cl,%eax
 701:   c3                      ret

0000000000000710 futility_move_count3:
 710:   f3 0f 1e fa             endbr64
 714:   0f af f6                imul   %esi,%esi
 717:   89 f9                   mov    %edi,%ecx
 719:   83 f1 01                xor    $0x1,%ecx
 71c:   8d 46 03                lea    0x3(%rsi),%eax
 71f:   d3 f8                   sar    %cl,%eax
 721:   c3                      ret

**Comment by dubslow** [comment]
tfw gcc sux

**Comment by cj5716** [comment]
Btw I have started LTC: https://tests.stockfishchess.org/tests/view/6426e232db43ab2ba6f9f1dc
Just forgot to update here

**Comment by cj5716** [comment]
It just failed LTC with yellow https://tests.stockfishchess.org/tests/view/6426e232db43ab2ba6f9f1dc
It never dropped to negative elo much, just didn't gain enough to pass

**Comment by dubslow** [comment]
Exactly as we all predicted lol

**Comment by mstembera** [comment]
LTC is less sensitive to speedups.  Also upon further inspection a significant portion of the fishtest workers are in fact running clang which interferes w/ the test since those produce identical assembly but gcc does not.

**Comment by maximmasiutin** [comment]
@mstembera oops, it turned out to be that was me who screwed the outcomes by adding these clang workers, sorry :-(

I was the person who supported this patch because I thought we should write such a C code that even such compiler as GCC could produce efficient assembly out of it, and @cj5716 demonstrated exactly that.

**Comment by mstembera** [comment]
@maximmasiutin Your rate of CPU contribution is amazing!  Ty!
https://tests.stockfishchess.org/users
```maximmasiutin 	2 minutes ago 	47897 	72961 	5825726```

**Comment by cj5716** [comment]
> @mstembera - interesting. How to ask the site to choose compilers? clang 16 with "-O3" produces the following results:
> 
> ```
> 0000000000000650  futility_move_count:
>  650:   89 f9                   mov    %edi,%ecx
>  652:   0f af f6                imul   %esi,%esi
>  655:   8d 46 03                lea    0x3(%rsi),%eax
>  658:   80 f1 01                xor    $0x1,%cl
>  65b:   d3 e8                   shr    %cl,%eax
>  65d:   c3                      ret
> 
> 
> 0000000000000660 futility_move_count2
>  660:   89 f9                   mov    %edi,%ecx
>  662:   0f af f6                imul   %esi,%esi
>  665:   8d 46 03                lea    0x3(%rsi),%eax
>  668:   80 f1 01                xor    $0x1,%cl
>  66b:   d3 e8                   shr    %cl,%eax
>  66d:   c3                      ret
> 
> 
> 0000000000000670 futility_move_count3:
>  670:   89 f9                   mov    %edi,%ecx
>  672:   0f af f6                imul   %esi,%esi
>  675:   8d 46 03                lea    0x3(%rsi),%eax
>  678:   80 f1 01                xor    $0x1,%cl
>  67b:   d3 e8                   shr    %cl,%eax
>  67d:   c3                      ret
> ```
> 
> The source code is the following:
> 
> ```
> int futility_move_count(bool improving, int depth) {
>     return improving ? (3 + depth * depth)
>                      : (3 + depth * depth) / 2;
> }
> 
> int futility_move_count2(bool improving, int depth) {
>     return (3 + depth * depth) >> (improving ? 0 : 1);
> }
> 
> int futility_move_count3(bool improving, int depth) {
>     return (3 + depth * depth) >> (!improving);
> }
> 
> int main(int argc, char *argv[])
> {
>   return
>    futility_move_count(argc, argc) +
>    futility_move_count2(argc, argc) +
>    futility_move_count3(argc, argc);
> }
> ```

question, @maximmasiutin on godbolt with clang 16.0.0 and -O3 it compiles differently. I get
image

**Comment by maximmasiutin** [comment]
> question, @maximmasiutin on godbolt with clang 16.0.0 and -O3 it compiles differently. I get image

What is the point of using godbold if you just can take [the source ](https://github.com/official-stockfish/Stockfish/pull/4477#issuecomment-1492582193 )and compile it locally and run objdump on it? You can do it in a script. I could have used online services to debug regular expressions, but for the real stuff like CPU instructions I would not trust the online services :-)

**Comment by maximmasiutin** [comment]
@cj5716 - I support your bitshift patch in the form of `>> (improving ? 0 : 1);` because if a compiler cannot produce good enough code (like the default gcc that we use for Stockfish) we should help, and the version with `improving ? 0 : 1` produces best balance in terms of readability for humans (protects from further errors if improving will be replaced from bool to int) and the produced code.

**Comment by vondele** [comment]
since it didn't pass LTC, the code is not a simplification, and gets optimized depending on the compiler, I would skip merging of this one.

**Comment by pcordes** [comment]
Just for the record:

> question, @maximmasiutin on godbolt with clang 16.0.0 and -O3 it compiles differently. I get image

That's not what I get on Godbolt with clang 16.0.0 or GCC12.2.0 with `-O3`.  https://godbolt.org/z/fh1a55ofn shows asm that matches Maxim's local GCC and clang, with none of them being branchy.  (I switched Godbolt to AT&T syntax to match Maxim's disassembly that left out `-Mintel` from the objdump command line.)

His claims about Godbolt giving worse asm are unfounded; Godbolt uses builds of actual GCC and clang, configured without the usual distro defaults of `-fPIE -fstack-protector-strong`, so if you want to reproduce local output exactly, use those options with a compiler version that matches your local one.  (We [discussed this on Stack Overflow](https://stackoverflow.com/questions/76027635/extract-four-xmm-registers-from-one-zmm-register-and-vice-versa-using-c-intrinsi/76028079?noredirect=1#comment134092485_76028079), and Maxim figures he must have made a mistake using Godbolt to reach that wrong conclusion.)

The real utility of Godbolt is having access to many compiler versions for many different ISAs, including nightly builds of GCC and clang updated every day.  Also the nice filtering of the asm, and updating as you modify the source to see how stuff compiles.  You can install Matt Godbolt's Compiler Explorer tools locally (since they're open source) if you prefer using your local CPU to do the work, but then you're responsible for updating the set of compilers.

**Pull request closed without merging**